lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/cache/EvictableLoadingCache.java

```diff
@@ -0,0 +1,369 @@
+        }
+
+        return new EvictableLoadingCache<>(
+                EvictableCache.buildWith(tokenCache),
+                dataCache.build(new TokenClassLoader<>(cacheLoader)));
+    }
+
+    private final EvictableCache<K, Token<K>> tokensCache;
+    private final LoadingCache<Token<K>, V> dataCache;
+
+    private EvictableLoadingCache(EvictableCache<K, Token<K>> tokensCache, LoadingCache<Token<K>, V> dataCache)
+    {
+        this.tokensCache = tokensCache;
+        this.dataCache = dataCache;
+    }
+
+    @Override
+    public V get(K key)
+            throws ExecutionException
+    {
+        Token<K> token = tokensCache.get(key, () -> new Token<>(key));
+        return dataCache.get(token);
+    }
+
+    @Override
+    public ImmutableMap<K, V> getAll(Iterable<? extends K> keys)
+            throws ExecutionException
+    {
+        BiMap<K, Token<K>> keyToToken = HashBiMap.create();
+        for (K key : keys) {
+            // This is not bulk, but is fast local operation
+            keyToToken.put(key, tokensCache.get(key, () -> new Token<>(key)));
+        }
+
+        Map<Token<K>, V> values = dataCache.getAll(keyToToken.values());
+
+        BiMap<Token<K>, K> tokenToKey = keyToToken.inverse();
+        ImmutableMap.Builder<K, V> result = ImmutableMap.builder();
+        for (Map.Entry<Token<K>, V> entry : values.entrySet()) {
+            Token<K> token = entry.getKey();
+
+            // While token.getKey() returns equal key, a caller may expect us to maintain key identity, in case equal keys are still distinguishable.
+            K key = tokenToKey.get(token);
+            checkState(key != null, "No key found for %s in %s when loading %s", token, tokenToKey, keys);
+
+            result.put(key, entry.getValue());
+        }
+        return result.buildOrThrow();
+    }
+
+    @Override
+    public V getIfPresent(Object key)
+    {
+        Token<K> token = tokensCache.getIfPresent(key);
+        if (token == null) {
+            return null;
+        }
+        return dataCache.getIfPresent(token);
+    }
+
+    @Override
+    public void invalidate(Object key)
+    {
+        Token<K> token = tokensCache.getIfPresent(key);
+        // Invalidate irrespective of whether we got a token, to invalidate result of ongoing token load (instantiation).
+        tokensCache.invalidate(key);
+        if (token != null) {
+            dataCache.invalidate(token);
+        }
+    }
+
+    @Override
+    public void invalidateAll(Iterable<?> keys)
+    {
+        tokensCache.invalidateAll(keys);
+
+        ImmutableMap<K, Token<K>> tokens = tokensCache.getAllPresent(keys);
+        tokensCache.invalidateAll(keys);
+        // This is racy with respect to concurrent cache laoding. dataCache is not guaranteed not to have entries for keys after invalidateAll(keys) returns.
+        // Note that dataCache invalidation is not necessary for correctness. Still one can expect that cache invalidation releases memory.
```

Comment:
I would add a general comment to a class to explain how `tokensCache` and `dataCache` work together. Then we could skip comment like this or at least we could simplify it a lot. Now I was confused seeing that big comment as I understood the relation between caches but then I was thinking that still is something missing while it is not.

Reply:
Added general class javadoc
```java
/**
 * A {@link LoadingCache} implementation similar to ones produced by {@link CacheBuilder#build(CacheLoader)},
 * but one that does not exhibit <a href="https://github.com/google/guava/issues/1881">Guava issue #1881</a>:
 * a cache inspection with {@link #getIfPresent(Object)} or {@link #get(Object)} is guaranteed to return fresh
 * state after {@link #invalidate(Object)}, {@link #invalidateAll(Iterable)} or {@link #invalidateAll()} were called.
 */
```

also added some impl comment at the fields declaration

```java
    // Token<K> is a freshness marker. tokenCache keeps the fresh token for given key.
    // Cache invalidation correctness (read-after-invalidate consistency) relies on tokenCache
    // invalidation.
    private final EvictableCache<K, Token<K>> tokensCache;

    // dataCache keeps data for tokens, potentially including stale entries.
    // dataCache invalidation is opportunistic. It's necessary for releasing memory, but
    // is *not* necessary for correctness (read-after-invalidate consistency).
    private final LoadingCache<Token<K>, V> dataCache;
```

Reply:
Thanks. So now comments in invalide* methods can be simplified.

Reply:
prefer to keep as is